### PR TITLE
Code improvement

### PR DESCRIPTION
**Pull request by blasvicco**
### Change log
- 2018-09-30 Adding test file.
- 2018-09-30 Adding `Request` object.
- 2018-09-26 Using the small Docker image Alpine instead of Debian.
- 2018-09-26 Using docker-compose to build/up/down the image.
- 2018-09-26 Grouping project files inside the app folder.

### Motivation
Creating the first test file.
Reducing code lines from `BitfinexAPI.hpp` and cleaning it moving the `curl handlers` to the `Request` class.
Using a small docker image.
Using docker-compose to sync the app folder and its content between the docker image and the host machine.
Avoiding clone the project inside the docker image.
Separating the project file from the docker files.

### TODO
Test the `post` action.

### CODE REVIEW

**Review by MMquant — APPROVED** [review]
Looks good!

File: app/include/bfx-api-cpp/BitfinexAPI.hpp [hunk 1]
```diff
@@ -35,21 +35,15 @@
...
-#include <cryptopp/osrng.h>
-
 // internal jsonutils
 #include "jsonutils.hpp"
 
 // internal error
 #include "error.hpp"
 
+// internal TRequest
+#include "TRequest.hpp"
```
Comment (MMquant) [inline app/include/bfx-api-cpp/BitfinexAPI.hpp]: I though I'm clever but now I see I'm not. I can't figure out what means `T` in `TRequest.hpp` :). I would rename it to "two word name"  `T....Request.hpp`.

File: app/include/bfx-api-cpp/TRequest.hpp [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+
+      void setHeader(map<string, string> inHeader) {
+        header = inHeader;
+      }
+
+    private:
+      ////////////////////////////////////////////////////////////////////////
+      // Private properties
+      ////////////////////////////////////////////////////////////////////////
+      string endpoint = "";
```
Comment (MMquant) [inline app/include/bfx-api-cpp/TRequest.hpp]: I think string initialization `string endpoint = "";` is unnecessary. Simple less verbose `string endpoint;` should do the same.

> (1) empty string constructor (default constructor)
    Constructs an empty string, with a length of zero characters.

File: app/include/bfx-api-cpp/TRequest.hpp [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+            new HashFilter(hmac, new StringSink(mac))
+          );
+          StringSource ss2(mac, true, new HexEncoder(new StringSink(digest)));
+          transform(digest.cbegin(), digest.cend(), digest.begin(), ::tolower);
+      };
+
+      void setupHeader() {
+        curlHeader = nullptr;
+        map<string, string>::iterator it;
+        for (it = header.begin(); it != header.end(); it++) {
```
Comment (MMquant) [inline app/include/bfx-api-cpp/TRequest.hpp]: Couldn't this code be simplified such ...
delete `map<string, string>::iterator it;`
change `for (auto it = header.begin(); it != header.end(); ++it) {`
?
But it's up to you if you prefer more verbose way. `auto` keyword comes handy in these cases.

File: app/include/bfx-api-cpp/BitfinexAPI.hpp (hunk 1 above)
Comment (blasvicco) [reply app/include/bfx-api-cpp/BitfinexAPI.hpp]: Sorry for it. It is just my way to define `class`, so then we can instance with the name of the object. So in this case it is `TRequest` so then we can do:
```c++
TRequest Request = new TRequest();
Request.get(...);
```
So in my mind `T` is `Template` like a way to call the recipe of abstraction (class) of the object (instance). But of course we can name it as you prefer.

Comment (MMquant) [reply app/include/bfx-api-cpp/BitfinexAPI.hpp]: I was looking in `TRequests.hpp` and I would rename all occurrences of `TRequest` to `HTTPRequest` which will improve code clarity.

**Review by MMquant — APPROVED** [review]

**Review by MMquant — APPROVED** [review]

**Review by MMquant — APPROVED** [review]

**Review by MMquant — APPROVED** [review]

File: app/include/bfx-api-cpp/BitfinexAPI.hpp (hunk 1 above)
Comment (blasvicco) [reply app/include/bfx-api-cpp/BitfinexAPI.hpp]: @MMquant sorry for the delay. I will change as soon as I have a moment.

Comment (blasvicco) [reply app/include/bfx-api-cpp/BitfinexAPI.hpp]: I'll change to `HTTPRequest` even if the request is not limited to the HTTP protocol.

**Review by MMquant — APPROVED** [review]

**Pull request merged**